src/utilities.py

```diff
@@ -0,0 +1,183 @@
+        """
+
+        @functools.wraps(func)
+        def fn_with_retry(*args, **kwargs) -> ReturnT:
+            """Wrap the function with retries."""
+            remain_tries, current_delay = tries, delay
+
+            for _ in range(tries):
+                try:
+                    return func(*args, **kwargs)
+                # Error caught is set by the input of the function.
+                except exception as err:  # pylint: disable=broad-exception-caught
+                    remain_tries -= 1
+
+                    if remain_tries == 0:
+                        if local_logger is not None:
+                            local_logger.exception("Retry limit of %s exceed: %s", tries, err)
+                        raise
+
+                    if local_logger is not None:
+                        local_logger.warning(
+                            "Retrying error in %s seconds: %s", current_delay, err
+                        )
+                        local_logger.debug("Error to be retried:", stack_info=True)
+
+                    time.sleep(current_delay)
+
+                    current_delay *= backoff
+
+                    if max_delay is not None:
+                        current_delay = min(current_delay, max_delay)
+
+            raise RuntimeError("Unreachable code of retry logic.")
+
+        return fn_with_retry
+
+    return retry_decorator
+
+
+def secure_run_subprocess(cmd: Sequence[str], **kwargs) -> subprocess.CompletedProcess[bytes]:
+    """Run command in subprocess according to security recommendations.
+
+    The argument `shell` is set to `False` for security reasons.
+
+    The argument `check` is set to `False`, therefore, CalledProcessError will not be raised.
+    Errors are handled by the caller by checking the exit code.
+
+    Args:
+        cmd: Command in a list.
+        kwargs: Additional keyword arguments for the `subprocess.run` call.
+
+    Returns:
+        Object representing the completed process. The outputs subprocess can accessed.
+    """
+    logger.info("Executing command %s", cmd)
+    result = subprocess.run(  # nosec B603
+        cmd,
+        capture_output=True,
+        shell=False,
+        check=False,
+        # Disable type check due to the support for unpacking arguments in mypy is experimental.
+        **kwargs,  # type: ignore
+    )
+    logger.debug("Command %s returns: %s", cmd, result.stdout)
+    return result
+
+
+def execute_command(cmd: Sequence[str], check_exit: bool = True, **kwargs) -> str:
+    """Execute a command on a subprocess.
+
+    The command is executed with `subprocess.run`, additional arguments can be passed to it as
+    keyword arguments. The following arguments to `subprocess.run` should not be set:
+    `capture_output`, `shell`, `check`. As those arguments are used by this function.
+
+    Args:
+        cmd: Command in a list.
+        check_exit: Whether to check for non-zero exit code and raise exceptions.
```

Comment:
I couldn't find where this check_exit is used :(

Reply:
There is usage in the `charm.py` under the `_install_deps` method and `_kernel_upgrade` method.